src/gt4py/caching.py

```diff
@@ -0,0 +1,424 @@
+    from gt4py.stencil_builder import StencilBuilder
+
+
+class CachingStrategy(abc.ABC):
+    name: str
+
+    def __init__(self, builder: "StencilBuilder"):
+        self.builder = builder
+
+    @property
+    @abc.abstractmethod
+    def root_path(self) -> pathlib.Path:
+        """
+        Calculate and if necessary create the root file path for stencil generation operations.
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def backend_root_path(self) -> pathlib.Path:
+        """
+        Calculate and if necessary create the caching base path for the backend being used.
+
+        Avoid trying to create the root_path automatically.
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def cache_info_path(self) -> Optional[pathlib.Path]:
+        """
+        Calculate the file path where caching info for the current process should be stored.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def generate_cache_info(self) -> Dict[str, Any]:
+        """
+        Generate the cache info dict.
+
+        Backend specific additions can be added via a hook propery on the backend instance.
+        Override :py:meth:`gt4py.backend.base.Backend.extra_cache_info` to store extra
+        info.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def update_cache_info(self) -> None:
+        """
+        Regenerate cache info and update the cache info file.
+
+        Parameters
+        ----------
+        extra_cache_info:
+            Additional info to be written
+
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def is_cache_info_consistent(
+        self, cache_info: Dict[str, Any], *, validate_hash: bool,
+    ) -> bool:
+        """
+        Validate the passed cache info by comparing it to the stored data.
+
+        A backend can cause additional keys to be validated by overriding
+        :py:meth:`gt4py.backend.base.Backend.extra_cache_validate`.  These keys must also be
+        present in it's overload of :py:meth:`gt4py.backend.base.Backend.extra_cache_info`.
+
+        Parameters
+        ----------
+        cache_info:
+            The cache info to be compared to file
+
+        validate_hash:
+            Should the stencil fingerprint and source code hash be validated or not
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def is_cache_info_available_and_consistent(self, *, validate_hash: bool) -> bool:
+        """
+        Check if the cache can be read and is consistent.
+
+        If the cache info file can be read, try to validate it against the current info
+        as returned by py:meth:`generate_cache_info`.
+
+        Parameters
+        ----------
+        validate_hash:
+            Should the stencil fingerprint and source code hash be validated or not
+        """
+        raise NotImplementedError
```

Comment:
Can you elaborate on why are these two methods needed? They seem a bit redundant since one looks like a stricter version of the other. Wouldn't it be simpler to define only one method with an extra parameter (e.g., `ignore_unavailable`, `strict_validation`, ...) to differentiate between the lenient and the strict mode?